core/src/main/java/org/everit/json/schema/NumberSchema.java

```diff
@@ -109,6 +111,8 @@ public static Builder builder() {
 
     private final boolean requiresInteger;
 
+    private final List<ValidationException> validationExceptions;
```

Comment:
This shouldn't be a member field, because this way the `NumberSchema` instance won't be thread-safe. I suggest handling it as a local variable in the `validate()` method and change the signatures of the `check*` methods to return an `Optional<ValidationException>` (or a `List<ValidationException>` if there are possible multiple errors checked). 

You can work from [StringSchema](https://github.com/everit-org/json-schema/blob/master/core/src/main/java/org/everit/json/schema/StringSchema.java#L151) as an example.

Reply:
@erosb thank you for feedback, I did not know it should be thread safe. do you think that with the approach like StringSchema you generate to way many unnecessary objects? For example, you create 3 ArrayLists even though it is possible they will never used and I did not count call to addAll, which does copy of arrays.

Reply:
Correctness first. I wouldn't go into too much micro-optimizations here. At least not as long as it isn't verified with a perf. test that this is a bottleneck.
This entire aspect should be refactored at some time anyway (I would like to make it configurable to collect all errors or fail immediately with the first one), so I really don't see the point in putting too much effort into reducing the GC load here.

Reply:
1. create list before hand and pass it as argument to check* (save 2 list creations)
2. (A bit ugly) create function:
```
    private List<ValidationException> getList(List<ValidationException> validationExceptions,
                                              ValidationException validationException) {
        if (validationExceptions == null) {
            validationExceptions = new ArrayList<>();
        }
        validationExceptions.add(validationException);
        return validationExceptions;
    }
```

Once you encountered an error in the check*, you call that function by passing listA and then assign the result of that function back to the listA, and so forth.

Reply:
If anything, then it makes more sense to optimize for the initial size of the list. `new ArrayList<>()` allocates space for 10 objects. In case a `check*` method can find _at most_ 2 errors, you can create the list with `new ArrayList<>(2)` .

Reply:
Okay, I see your point, I did it a little bit different from StringSchema. Please, check it out.